### PR TITLE
Add compilation cache in react native CI

### PR DESCRIPTION
**Pull request by mszhanyi**
### Description
1. Replacing jobs with stages for better debugging and maintainance
2. Added compilation cache to accelerate the workflow.
3. Splited building protobuf and major code as 2 tasks

### Motivation and Context
Reduced compilation time about one hour.
test run:
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=943695&view=logs&j=de302ec2-2305-57e0-e8c6-cd89c569f2a3&t=8b360243-7783-51da-8079-2304089d3d1d

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml [hunk 1]
```diff
@@ -20,302 +20,338 @@ parameters:
...
+    - script: |
+        python3 -m pip install -q flatbuffers
+      displayName: Install python modules
+
+    - template: mac-build-step-with-cache.yml
+      parameters:
+        TODAY: $(TODAY)
+        AdditionalKey: ' protobuf | "$(Agent.OS)" | $(Build.SourcesDirectory)/cmake/deps.txt, $(Build.SourcesDirectory)/tools/ci_build/github/linux/docker/inference/x64/python/cpu/scripts/install_protobuf.sh'
+        CACHE_DIR: $(PROTO_CACHE_DIR)
+        ChangeEveryCommit: false
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml:85]: I'm not too familiar with ccache. could you explain what is being cached in this step and the next one, and why this one has ChangeEveryCommit set to false?

Comment (mszhanyi) [reply tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml:85]: It‘s used to save storage cost. 
For pipeline cache, new cache is generated with the new key.
Protobuf isn‘t changed frequently and we can check the change by 2 files.  It's a waste to generate new cache in every commit.
Onnxruntime code is updated in every commit, so the Build.SouceVersion(commitid) is part of the key.

Comment (edgchen1) [reply tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml:85]: thanks for the explanation and link.
the protobuf caching makes sense.
for the onnxruntime build, why do we want to have a cache entry for each commit? how often do we expect to build at the same commit?

File: tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml [hunk 2]
```diff
@@ -13,30 +13,32 @@ parameters:
...
 
+- name: ChangeEveryCommit
+  type: boolean
+  default: true
+
 steps:
   - script: |
-      brew install ccache
+      which ccache || brew install ccache
       echo "##vso[task.prependpath]/usr/local/opt/ccache/libexec"
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml:23]: so if there is an existing ccache, we assume it is a brew installation of ccache.
maybe something like `brew list ccache &> /dev/null || brew install ccache` above then?

[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml:23]: It‘s a big risk that there's any unknown installation in build machine.

File: tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml (hunk 1 above)
Comment (mszhanyi) [reply tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml:85]: It's too complicated to know which commit would generate different C/C++ outputs.
besides changes in c/c++ files, other files could change compiler options.
`a cache entry for each commit` is to used generate cache for next builds and it's a safe way to keep the hit rate.

With restore keys, the pipeline would download cache generated by previous runs even they aren't same commit.
`This is useful if the pipeline is unable to find an exact match but wants to use a partial cache hit instead.`
https://learn.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#restore-keys.

Other useful info could be checked out in [cache isolation and security]
(https://learn.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#cache-isolation-and-security)
To save storage cost,  I'm trying to force merge branch to read the cache only generated in main branch in #15330

Comment (edgchen1) [reply tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml:85]: I see, thanks.

**Review by edgchen1 — APPROVED** [review]

File: tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml:23]: Unknow installation pose security risk.
https://en.wikipedia.org/wiki/2020_United_States_federal_government_data_breach
`On May 12, 2021, President Biden issued an [Executive Order on Improving the Nations Cybersecurity](https://www.whitehouse.gov/briefing-room/presidential-actions/2021/05/12/executive-order-on-improving-the-nations-cybersecurity/).  It put compliance requirements on build system and software development lifecycle of the software provided to the U.S. government. At Microsoft, all customer facing cloud services and all the latest released versions of our on-prem products will need to conform with the Executive Order.`

**Review by snnn — APPROVED** [review]

File: tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (edgchen1) [reply tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml:23]: I got the part that unknown installations are a security risk. I didn't understand how it's being mitigated. and aren't we using hosted Mac agents which have known installed software?

File: tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml [hunk 3]
```diff
@@ -13,30 +13,32 @@ parameters:
 - name: CACHE_DIR
   type: string
 
+- name: ChangeEveryCommit
```
[the remark this answers is not in the document]
Comment (mszhanyi) [reply tools/ci_build/github/azure-pipelines/templates/mac-build-step-with-cache.yml:16]: Updated in #15440

**Pull request merged**